File: exercism/go/protein-translation/protein_translation.go
package protein

import "errors"

var protainMap = map[string]string{
	"AUG": "Methionine",
	"UUU": "Phenylalanine",
	"UUC": "Phenylalanine",
	"UUA": "Leucine", "UUG": "Leucine",
	"UCU": "Serine", "UCC": "Serine", "UCA": "Serine", "UCG": "Serine",
	"UAU": "Tyrosine", "UAC": "Tyrosine",
	"UGU": "Cysteine", "UGC": "Cysteine",
	"UGG": "Tryptophan",
	"UAA": "STOP", "UAG": "STOP", "UGA": "STOP",
}
var ErrInvalidBase = errors.New("ErrInvalidBase")
var ErrStop = errors.New("ErrStop")

func FromRNA(rna string) ([]string, error) {
	var result []string

	for n := 0; n <= len(rna)-3; n += 3 {
		protain, ok := FromCodon(rna[n : n+3])

		if ok == ErrInvalidBase {
			return result, ErrInvalidBase
		} else if ok == ErrStop {
			return result, nil
		}
		result = append(result, protain)
	}
	return result, nil
}

func FromCodon(codon string) (string, error) {
	protain, ok := protainMap[codon]
	if !ok {
		return "", ErrInvalidBase
	} else if protain == "STOP" {
		return "", ErrStop
	} else {
		return protain, nil
	}
}

File: exercism/python/protein-translation/protein_translation.py
"""
Protein Translation
"""

PROTAIN_MAP = {
  "Methionine": ["AUG"],
  "Phenylalanine": ["UUU", "UUC"],
  "Leucine": ["UUA", "UUG"],
  "Serine": ["UCU", "UCC", "UCA", "UCG"],
  "Tyrosine": ["UAU", "UAC"],
  "Cysteine": ["UGU", "UGC"],
  "Tryptophan": ["UGG"],
  "STOP": ["UAA", "UAG", "UGA"]
}


def proteins(strand):
    """
    Translate RNA sequences into proteins.
    """
    result = []
    # used to split the condons
    number = 3
    codons = [strand[i:i+number] for i in range(0, len(strand), number)]

    for codon in codons:
        if codon in PROTAIN_MAP['STOP']:
            break
        for protein, values in PROTAIN_MAP.items():
            if codon in values:
                result.append(protein)

    return result
